### DISCUSSION TITLE
AEIP - add a code property to the token standard

### DISCUSSION THREAD

**Discussion by Bantarus** [archethic-foundation/aeip#10] [Ideas]
Hello Archers,

I was looking at the token standard you are defining.
I like the choice of agnostic approach.

Could it means we could have a  "smart" token that embbed some intelligence throught a "code" property ? 

This "code" would accept the same language as the one you are defining for the smart contract writing.

The goals would be to have a token generated with an inalterable feature instead of having this feature managed by an ordinary smart contract.

For examples : I could like to have a token to be splitted in two tokens every time it is transferred to another chain.

Or maybe i would like a token that is the only one that contain the function to encrypt or decrypt some secret and i could send this token to another address to allow the owner to decrypt this secret.

Another use case would be my token could be the base intelligence of an basic AI that could be extended by an ordinary smart contract.

This could be a way to transfert not only images, money etc... but also intelligence. We could make it so that intelligence become a transferrable value.

What do you think about this ? 
Does it make any sense to you ?
Or maybe you are already handling this in some mecanism i am not aware ?

Curious to have some insight from the community and the core team.

See you.

Bantarus

**Comment by ghost** [comment]
Hello.

Thanks for your comment.

So indeed a token could have some smart logic through the code property, because any transaction can become a smart contract. 
For example, in the case of a crowdsale/ICO you could define a fungible token and still mention the rules how to distribute again UCO payments—still in the same transaction.

Also, we are working on another improvement for UTXO which are to make them smarter through the use of hooks.
Because we are using the UTXO model, once a token is sent, the owner don't have control anymore on it. 
Hence, we would like to give the capability to define a behavior before to send it to allow logic when this token will be sent or spent.

Following this idea, the token creation would accept a field to describe the hooks to take the tokens smarter and trigger some code when they are used.

So stay tuned, for the next AEIPs 😉

**Reply by Bantarus** [reply]
Hello Samuel,

Thanks for your clear answer.

Also @Neylix answered to me by directing me towards the issue : https://github.com/archethic-foundation/archethic-docs/issues/7#issue-1042297569

The way you are thinking to implement it at the transaction level make even more sense.

Bravo l'équipe ! 
I still can't find a complaint to make about Archethic, awesome technology, i'm very impressed so far 👌 
 
See you 🙂
